crates/bounded_planet_core/src/unit_selection.rs

```diff
@@ -0,0 +1,207 @@
+    cursor_moved: Res<Events<CursorMoved>>,
+
+    query_pipeline: Res<QueryPipeline>,
+    colliders: Res<ColliderSet>,
+    mouse_button_inputs: Res<Input<MouseButton>>,
+
+    windows: Res<Windows>,
+
+    mut camera_query: Query<(&Camera, &GlobalTransform)>,
+    mut colliders_query: Query<(Entity, &ColliderHandleComponent)>,
+) {
+    // Save the current cursor position every time it moves. If this is only done when the mouse button is pressed
+    // then you'll often read no cursor moved events since it only holds the event for 2 ticks max.
+    if let Some(CursorMoved { position, .. }) = state.cursor_moved_reader.latest(&cursor_moved) {
+        state.cursor_position = *position;
+    }
+    
+    if mouse_button_inputs.just_pressed(MouseButton::Left) {
+        // If the user clicks at all, clear the selection state. Aka, deselect whatever is selected
+        selection_state.swap_and_clear_current();
+        
+        // TODO: properly figure out _which_ camera was intended to be used
+        // Search for a camera with the default 3d camera name, which should be the "main" camera in a 3d game
+        let mut camera_query_iter = camera_query.iter();
+        let (camera, camera_transform) = camera_query_iter.into_iter()
+            .find(|&(camera, _)| {
+                // If Some(name), return name == base::camera::CAMERA3D. If None, return false
+                camera.name.as_ref().map_or(false, |name| name == base::camera::CAMERA3D)
+            })
+            .expect("Unable to find a main camera! Do you not have a Camera3dComponents in use?");
+
+        let window = windows.get(camera.window)
+            .expect("Unable to find the window of the 3d camera! Are you running headless yet trying to do screen-based unit selection?");
+        let screen_size = Vec2::new(window.width as f32, window.height as f32);
+
+        // Turn the cursor position on the screen into a position in world space (but still on the camera plane)
+        let point = math::screen_to_world(
+            state.cursor_position,
+            screen_size,
+            *camera_transform.value(),
+            camera.projection_matrix,
+        );
+
+        // Construct a direction vector starting from the camera's location in world space, passing through the cursor position in world space
+        let direction = point - camera_transform.translation();
+        // Construct the ray starting at the screen point in world space, heading in the computed direction
+        let ray = Ray::new(point.into_rapier(), direction.into_rapier());
+
+        if let Some((hit_collider, _, _)) = query_pipeline.cast_ray(&colliders, &ray, std::f32::MAX) {
+            // The raycast hit something!
+
+            // Search for an entity with the same collider handle as the collider we hit
+            let mut colliders_query_iter = colliders_query.iter();
+            let query_search = colliders_query_iter.into_iter()
```

Comment:
We definitely need a better way to find this. If Bevy-Rapier doesn't do it already (e.g. I would expect the result returned from `cast_ray` to have an `Entity` reference at least) we'll need to build an index mapping between `Entity<->Collider Handler` (I'm ok with a todo for that).

Reply:
#67 is now open